File: src/cli/sed.go
package cli

import (
	"fmt"
	"strings"
)

// Sed wraps sed commands
type Sed struct {
	delimiter string
	flags     string
}

// NewSed creates a new Sed command builder
func NewSed() *Sed {
	return &Sed{
		delimiter: "/",
		flags:     "g",
	}
}

// Delimiter sets delimiter
func (s *Sed) Delimiter(value string) *Sed {
	s.delimiter = value
	return s
}

// Flags sets flags
func (s *Sed) Flags(value string) *Sed {
	s.flags = value
	return s
}

// Replace generates sed command to replace text in a file
func (s *Sed) Replace(path, search, replacement string) string {
	escapedSearch := escapeDelimiter(escapeSingleQuotes(search), s.delimiter)
	escapedReplacement := escapeDelimiter(escapeSingleQuotes(replacement), s.delimiter)
	expression := fmt.Sprintf("s%s%s%s%s%s", s.delimiter, escapedSearch, s.delimiter, escapedReplacement, s.delimiter+s.flags)
	return fmt.Sprintf("sed -i '%s' %s", expression, quote(path))
}

func escapeDelimiter(value, delimiter string) string {
	return strings.ReplaceAll(value, delimiter, "\\"+delimiter)
}

File: csharp-enva/CLI/Sed.cs
using System.Text.RegularExpressions;

namespace Enva.CLI;

public class Sed
{
    private string delimiter = "/";
    private string flags = "g";

    public static Sed NewSed()
    {
        return new Sed();
    }

    public Sed Delimiter(string value)
    {
        delimiter = value;
        return this;
    }

    public Sed Flags(string f)
    {
        flags = f;
        return this;
    }

    public string Replace(string path, string search, string replacement)
    {
        string escapedSearch = EscapeDelimiter(EscapeSingleQuotes(search), delimiter);
        string escapedReplacement = EscapeDelimiter(EscapeSingleQuotes(replacement), delimiter);
        string expression = $"s{delimiter}{escapedSearch}{delimiter}{escapedReplacement}{delimiter}{flags}";
        return $"sed -i '{expression}' {Quote(path)}";
    }

    private string EscapeDelimiter(string value, string delim)
    {
        return value.Replace(delim, "\\" + delim);
    }

    private string EscapeSingleQuotes(string value)
    {
        // Go: strings.ReplaceAll(value, "'", "'\"'\"'")
        return value.Replace("'", "'\"'\"'");
    }

    private string Quote(string s)
    {
        if (s.Contains(" ") || s.Contains("$") || s.Contains("'"))
        {
            return $"'{s.Replace("'", "'\"'\"'")}'";
        }
        return s;
    }
}
